### PR TITLE
Normalize hydro

### LINKED ISSUE #375: Add hydro inflow normalization
## Describe the feature you'd like to see

While reviewing PR https://github.com/pypsa-meets-africa/pypsa-africa/pull/366, the hydro production is about 40% off; there is the need to normalize the inflows to adhere to the average inflows as per declared in hydro_capacities.

### PR DESCRIPTION
**Pull request by davide-f**
# Closes #375 

## Changes proposed in this Pull Request
This PR aims at adding hydro inflows normalization options.
I think normalization may be done in three options:
- [x] no normalization
- [x] normalization by country: the runoff of hydro is normalized to match the energy totals for buses where hydro is installed
- [x] (maybe) normalization by whole region: to match the totals of the whole region under interest

Note: this PR has requested deep investigation of atlite and adding a feature that is currently pending: https://github.com/PyPSA/atlite/pull/245

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by davide-f** [comment]
@pz-max you can start have a look to the PR, the CI is working.
However, before merging there is the need to:
- [x] address inflows for countries with no existing hydro; with the current implementation the value may be put to 0.0 or errors may appear
- [ ] wait for comments from @euronion in atlite regarding the https://github.com/PyPSA/atlite/pull/245 this PR relies on. Currently we are working on a branch, though
- [x] test the workflow for large regions

File: envs/environment.yaml
Comment (pz-max) [inline envs/environment.yaml]: You mean "until ... is merged"

File: config.default.yaml
Comment (pz-max) [inline config.default.yaml]: can we add a comment with the untit here? 
is it: # m/s  ??

File: config.default.yaml
Comment (pz-max) [inline config.default.yaml:256]: Nice!

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py:348]: can we rename "plants" to "hydro_plants"

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: This filtering is a bit confusing/complex. :D I know it's compact but hard to maintain. Do you have ideas how to simplify it?

One idea: 
- put the list comprehension out of .loc statement
- ... maybe you have some more ideas

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py:375]: `if normalization == "eia":` is missing?

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: would it make sense to read in this snakemake input just below the __main__?

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: This will multiply hydro_prod_by_country by 21. Intended?

**Review by pz-max — APPROVED** [review]
Hi Davide, really cool! In summary:
- "eia" is missing for the normalization?
- if possible, some code could be simplified
But in general, if the code runs it's good and ready to merge.
See more in the review comments.

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py:375]: Currently yes, I tried to investigate its use using the pypsa-eur-like approach, however I couldnt' find the exact platform to have the data in the same format. I found the format, but while downloading the data a different dataformat was provided, hence I first used the hydro_capacities data that I think it is a nice-have anyawy
I can ask them.

@euronion , could you support us? I have tried to find the user interface, but from [this link described in build_hydro_profile of pypsa-eur](https://www.eia.gov/international/data/world#/?pa=000000000000000000000000000000g&c=1028i008006gg6168g80a4k000e0ag00gg0004g800ho00g8&ct=0&ug=8&tl_id=2-A&vs=INTL.33-12-ALB-BKWH.A&cy=2014&vo=0&v=H&start=2000&end=2016), I cannot find the way how to extract the data in the EIA format of pypsa-eur, as downloable from [link]( https://www.eia.gov/international/api/series_data/csv?location32=ruvvvvvfvtvnvv1vrvvvvfvvvvvvfvvvou20evvvvvvvvvvnvvvs0008&products32=000000000000000000000000000000g&frequency=A&unit=1&start=315532800000&end=1577836800000&generated=03-05-2022_10-09-49).
Could you describe how to do that? I would like to use the interface to create the EIA file for all world instead but in the attempts the link was not downloading the output as series_data (see the text of the last url)

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: We can do that, though I check the possible readability when generalizing this.
Currently I'm on a train with poor connection, hard to work on this

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: no. Example: [2] * 3 = [2 2 2].
It is a way to initialize a list with the same value in every element.
This is to create a eia-like dataframe where for every input year, the same hydro_capacities data are used

File: envs/environment.yaml
Comment (davide-f) [reply envs/environment.yaml]: better

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py:348]: No, "plants" is a field of the function hydro of atlite and in this dictionary we are preparing the inputs
We can propose a change in atlite to change such name, but I think it is clear as is, we can ask @euronion .
I think it is clear as is in atlite and if so, we should adhere to that

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: As long as it is main, I think we could keep using snakemake.{smt} as long as that parameter is used only once or twice in the code.
I think that initializing variables at the beginning that are used only once may create only large chuncks of variables at the beginning with scarce benefits, what do you think?

**Review by pz-max — APPROVED** [review]

File: scripts/build_renewable_profiles.py
Comment (pz-max) [reply scripts/build_renewable_profiles.py:348]: Ah ok! That's fine then. It would be just a "nice to have"

**Review by pz-max — APPROVED** [review]
@davide-f Seems this PR is ready to merge with only the `hydro_capacities` approach. Should we track the "eia" enhancement as a separate issue?

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py:375]: Just saw this issue: https://github.com/PyPSA/pypsa-eur/issues/324

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py]: Maybe two parentheses were too many, but I liked to move the list comprehension outside.
Done option 1.

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py:375]: Implemented a preliminry eia methodology based on pypsa-eur

**Comment by davide-f** [comment]
> @davide-f Seems this PR is ready to merge with only the `hydro_capacities` approach. Should we track the "eia" enhancement as a separate issue?

I've implemented a draft version, testing right now

**Comment by davide-f** [comment]
In the night, this PR is running for the continent. Fingers crossed

**Comment by davide-f** [comment]
As expected, bugs appear in the Africa analysis for regions where there are no existing hydro plants for example.
Some improvements are still needed

**Comment by davide-f** [comment]
This PR is improving.
By running the model for the entire Africa, this rasterio warning appears:
`WARNING:rasterio._env:CPLE_NotSupported in driver MEM does not support creation option PIXELTYPE`

I am ignoring it, but it is worth tracking it

**Comment by davide-f** [comment]
This PR is ready to be reviewed.
I've tested it with NG only and Africa. Even with the normalization we have a 20% error partially due to the inflows that lead to huge spills. Probably, this can be fixed by improving the atlite hydro modellign or using [lisflood](https://ec-jrc.github.io/lisflood/).
In the Africa case we obtain about 96TWh w.r.t. 120TWh.

Anyway, I discovered a bug on powerplantmatching: the CI country may not be well recognized because in some databases punctuation is not addressed. I have to create an issue on that.

File: notebooks/validation/hydro_generation_validation.ipynb
Comment (pz-max) [inline notebooks/validation/hydro_generation_validation.ipynb]: This is now GWh instead of TWh or not?

File: scripts/build_renewable_profiles.py
Comment (pz-max) [inline scripts/build_renewable_profiles.py:216]: Do you think we should move this and the below function in future in a dedicated rule as done in https://github.com/PyPSA/pypsa-eur/blob/master/scripts/build_hydro_profile.py

Comment (pz-max) [reply scripts/build_renewable_profiles.py:216]: Further, should we add a docstring with a link to the data? Should we also do that for `get_hydro_capacity_annual_hydro_generation()`?

**Review by pz-max — CHANGES_REQUESTED** [review]
Excellent work. Some remaining questions that could be addressed are described in the review. In general, I think the PR is ready to merge, let's just track some untackled issues/ or changes as an issue.

I think only the notebook needs minor changes .. convert GWh -> TWh

File: notebooks/validation/hydro_generation_validation.ipynb
Comment (davide-f) [reply notebooks/validation/hydro_generation_validation.ipynb]: The above is to avoid problems when snapshots do not contain 8760h.
In the case the snapshots do not contain such number of units, the first sum is not the total yearly production by plant, but a fraction. By using mean, we compute the hourly average whatever the length of the snapshot
Hence to compute the yearly total we can multiply such number by 8760.

Regarding the unit, yeah, I can change it into TWh :)

File: scripts/build_renewable_profiles.py
Comment (davide-f) [reply scripts/build_renewable_profiles.py:216]: > Do you think we should move this and the below function in future in a dedicated rule as done in https://github.com/PyPSA/pypsa-eur/blob/master/scripts/build_hydro_profile.py

Yeah, I was exactly thinking about that as well. We may think how we want the build_renewable profiles. Whether we want them to be split by resource or only hydro/all_the_rest.

> Further, should we add a docstring with a link to the data? Should we also do that for `get_hydro_capacity_annual_hydro_generation()`?

That's kind of duplicated with respect to the data (see link in the eia csv), but it can be useful

**Comment by pz-max** [comment]
@davide-f  ready to merge after your changes that you like to implement are added. Let me know :)

File: notebooks/validation/hydro_generation_validation.ipynb
Comment (davide-f) [reply notebooks/validation/hydro_generation_validation.ipynb]: Done

**Comment by davide-f** [comment]
> @davide-f ready to merge after your changes that you like to implement are added. Let me know :)

@pz-max, done :) please, give a final revise

**Comment by davide-f** [comment]
> > @davide-f ready to merge after your changes that you like to implement are added. Let me know :)
> 
> @pz-max, done :)

**Review by pz-max — APPROVED** [review]

**Pull request merged**
